distributed/shuffle/tests/test_shuffle.py

```diff
@@ -102,30 +108,247 @@ async def test_bad_disk(c, s, a, b):
     )
     out = dd.shuffle.shuffle(df, "x", shuffle="p2p")
     out = out.persist()
+    shuffle_id = await get_shuffle_id(s)
     while not a.extensions["shuffle"].shuffles:
         await asyncio.sleep(0.01)
     shutil.rmtree(a.local_directory)
 
     while not b.extensions["shuffle"].shuffles:
         await asyncio.sleep(0.01)
     shutil.rmtree(b.local_directory)
-    with pytest.raises(FileNotFoundError) as e:
+    with pytest.raises(
+        RuntimeError, match=f"shuffle_transfer failed .* {shuffle_id}"
+    ) as exc_info:
         out = await c.compute(out)
 
-    assert os.path.split(a.local_directory)[-1] in str(e.value) or os.path.split(
+    cause = exc_info.value.__cause__
+    assert isinstance(cause, FileNotFoundError)
+    assert os.path.split(a.local_directory)[-1] in str(cause) or os.path.split(
         b.local_directory
-    )[-1] in str(e.value)
+    )[-1] in str(cause)
 
     # clean_worker(a)  # TODO: clean up on exception
     # clean_worker(b)  # TODO: clean up on exception
     # clean_scheduler(s)
 
 
-@pytest.mark.skip
+async def wait_until_worker_has_tasks(
+    prefix: str, worker: str, count: int, scheduler: Scheduler, interval: float = 0.01
+) -> None:
+    ws = scheduler.workers[worker]
+    while (
+        len(
+            [
+                key
+                for key, ts in scheduler.tasks.items()
+                if prefix in key and ts.state == "memory" and ws in ts.who_has
+            ]
+        )
+        < count
+    ):
+        await asyncio.sleep(interval)
+
+
+async def wait_for_tasks_in_state(
+    prefix: str,
+    state: str,
+    count: int,
+    dask_worker: Worker | Scheduler,
+    interval: float = 0.01,
+) -> None:
+    tasks: Mapping[str, SchedulerTaskState | WorkerTaskState]
+
+    if isinstance(dask_worker, Worker):
+        tasks = dask_worker.state.tasks
+    elif isinstance(dask_worker, Scheduler):
+        tasks = dask_worker.tasks
+    else:
+        raise TypeError(dask_worker)
+
+    while (
+        len([key for key, ts in tasks.items() if prefix in key and ts.state == state])
+        < count
+    ):
+        await asyncio.sleep(interval)
+
+
+async def wait_for_cleanup(scheduler: Scheduler) -> None:
+    scheduler_extension = scheduler.extensions["shuffle"]
+    waits = []
+    for ev in scheduler_extension.removed_state_events.values():
+        waits.append(ev.wait())
+    await asyncio.gather(*waits)
+
+
+async def get_shuffle_id(scheduler: Scheduler) -> ShuffleId:
+    scheduler_extension = scheduler.extensions["shuffle"]
+    while not scheduler_extension.shuffle_ids():
+        await asyncio.sleep(0.01)
+    shuffle_ids = scheduler_extension.shuffle_ids()
+    assert len(shuffle_ids) == 1
+    return next(iter(shuffle_ids))
+
+
 @pytest.mark.slow
```

Comment:
This file skips 13 tests due to slowness, that's 25% of all test cases in this file. We should review 

1. Are the slow tests actually slow. This test, for instance, is only taking 0.8s on my machine. While arguably this is not fast, it is not slow either
2. Is there a systematic problem why these tests are slow? Are we sleeping often? Are we using too much data, etc.

Reply:
What's a good rule-of-thumb threshold for marking "slow" tests? I tend to go with tests approaching 1s, but I'll revise tests to double-check they are indeed slow.

Reply:
I've removed a flew flags, we're now down to 7 which seem to be genuinely slow.